Crash on null being returned by Firestore

App crashes when Firestore emits a null value, due to onError not being implemented.

```
FATAL EXCEPTION: main
Process: it.droidcon.android.duemiladiciotto, PID: 19016
io.reactivex.exceptions.OnErrorNotImplementedException: onNext called with null. Null values are generally not allowed in 2.x operators and sources.
	at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:704)
	at io.reactivex.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:701)
	at io.reactivex.internal.observers.LambdaObserver.onError(LambdaObserver.java:77)
	at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.checkTerminated(ObservableObserveOn.java:276)
	at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:172)
	at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:252)
	at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:109)
	at android.os.Handler.handleCallback(Handler.java:739)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.support.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:148)
	at android.support.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:465)
	at android.support.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:421)
	at android.support.test.espresso.base.UiControllerImpl.injectMotionEvent(UiControllerImpl.java:235)
	at android.support.test.espresso.action.MotionEvents.sendUp(MotionEvents.java:140)
	at android.support.test.espresso.action.MotionEvents.sendUp(MotionEvents.java:123)
	at android.support.test.espresso.action.Tap.sendSingleTap(Tap.java:170)
	at android.support.test.espresso.action.Tap.access$100(Tap.java:31)
	at android.support.test.espresso.action.Tap$1.sendTap(Tap.java:47)
	at android.support.test.espresso.action.GeneralClickAction.perform(GeneralClickAction.java:136)
	at android.support.test.espresso.ViewInteraction$SingleExecutionViewAction.perform(ViewInteraction.java:356)
	at android.support.test.espresso.ViewInteraction.doPerform(ViewInteraction.java:248)
	at android.support.test.espresso.ViewInteraction.access$100(ViewInteraction.java:63)
	at android.support.test.espresso.ViewInteraction$1.call(ViewInteraction.java:153)
	at android.support.test.espresso.ViewInteraction$1.call(ViewInteraction.java:150)
	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
	at android.os.Handler.handleCallback(Handler.java:739)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:135)
	at android.app.ActivityThread.main(ActivityThread.java:5538)
	at java.lang.reflect.Method.invoke(Native Method)
	at java.lang.reflect.Method.invoke(Method.java:372)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:960)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
Caused by: java.lang.NullPointerException: onNext called with null. Null values are generally not allowed in 2.x operators and sources.
	at io.reactivex.internal.operators.observable.ObservableCreate$CreateEmitter.onNext(ObservableCreate.java:63)
	at net.squanchy.service.firebase.FirestoreDbService$event$1$registration$1.onEvent(FirestoreDbService.kt:152)
	at net.squanchy.service.firebase.FirestoreDbService$event$1$registration$1.onEvent(FirestoreDbService.kt:20)
	at com.google.firebase.firestore.DocumentReference.zza(Unknown Source)
	at com.google.firebase.firestore.zzd.onEvent(Unknown Source)
	at com.google.android.gms.internal.zzeyn.zza(Unknown Source)
	at com.google.android.gms.internal.zzeyo.run(Unknown Source)
	... 26 more
```

--- a/app/src/main/java/net/squanchy/schedule/SchedulePageView.kt
+++ b/app/src/main/java/net/squanchy/schedule/SchedulePageView.kt
@@ -10,6 +10,7 @@ import io.reactivex.Observable
 import io.reactivex.android.schedulers.AndroidSchedulers
 import io.reactivex.disposables.CompositeDisposable
 import io.reactivex.functions.BiFunction
+import io.reactivex.schedulers.Schedulers
 import kotlinx.android.synthetic.main.view_page_schedule.view.*
 import net.squanchy.R
 import net.squanchy.analytics.Analytics
@@ -89,6 +90,7 @@ class SchedulePageView @JvmOverloads constructor(
     override fun startLoading() {
         subscriptions.add(
             Observable.combineLatest(service.schedule(), featureFlags.showEventRoomInSchedule.toObservable(), combineInPair())
+                .subscribeOn(Schedulers.io())
                 .observeOn(AndroidSchedulers.mainThread())
                 .subscribe(
                     { updateWith(it.first, it.second, ::onEventClicked) },

--- a/app/src/main/java/net/squanchy/service/firebase/FirestoreDbService.kt
+++ b/app/src/main/java/net/squanchy/service/firebase/FirestoreDbService.kt
@@ -2,9 +2,12 @@ package net.squanchy.service.firebase
 
 import com.google.android.gms.tasks.Task
 import com.google.firebase.firestore.DocumentReference
+import com.google.firebase.firestore.DocumentSnapshot
 import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.QuerySnapshot
 import io.reactivex.Completable
 import io.reactivex.Observable
+import io.reactivex.ObservableEmitter
 import net.squanchy.service.firebase.model.conferenceinfo.FirestoreConferenceInfo
 import net.squanchy.service.firebase.model.conferenceinfo.FirestoreVenue
 import net.squanchy.service.firebase.model.schedule.FirestoreEvent
@@ -15,8 +18,6 @@ import net.squanchy.service.firebase.model.schedule.FirestoreTrack
 import net.squanchy.service.firebase.model.twitter.FirestoreTweet
 import org.joda.time.DateTimeZone
 
-// TODO
-// val onCurrentThread = Executors.newSingleThreadExecutor { Thread.currentThread() }
 class FirestoreDbService(private val db: FirebaseFirestore) {
 
     fun scheduleView(): Observable<List<FirestoreSchedulePage>> {
@@ -24,14 +25,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
             val registration = view(VIEW_SCHEDULE)
                 .collection(COLLECTION_SCHEDULE_PAGES)
                 .orderBy(DAY_DATE_SORTING)
-                .addSnapshotListener { snapshot, exception ->
-                    if (exception != null) {
-                        subscriber.onError(exception)
-                        return@addSnapshotListener
-                    }
-
-                    subscriber.onNext(snapshot.documents.map { it.toObject(FirestoreSchedulePage::class.java) })
-                }
+                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -42,13 +36,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
             val registration = db.collection(SOCIAL_STREAM)
                 .document(VIEW_TWITTER)
                 .collection(COLLECTION_TWEETS)
-                .addSnapshotListener { snapshot, exception ->
-                    if (exception != null && subscriber.isDisposed.not()) {
-                        subscriber.onError(exception)
-                        return@addSnapshotListener
-                    }
-                    subscriber.onNext(snapshot.documents.map { it.toObject(FirestoreTweet::class.java) })
-                }
+                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -62,13 +50,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
         return Observable.create { subscriber ->
             val registration = db.collection(CONFERENCE_INFO)
                 .document(VENUE)
-                .addSnapshotListener { snapshot, exception ->
-                    if (exception != null && subscriber.isDisposed.not()) {
-                        subscriber.onError(exception)
-                        return@addSnapshotListener
-                    }
-                    subscriber.onNext(snapshot.toObject(FirestoreVenue::class.java))
-                }
+                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -78,13 +60,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
         return Observable.create { subscriber ->
             val registration = db.collection(CONFERENCE_INFO)
                 .document(CONFERENCE)
-                .addSnapshotListener { snapshot, exception ->
-                    if (exception != null && subscriber.isDisposed.not()) {
-                        subscriber.onError(exception)
-                        return@addSnapshotListener
-                    }
-                    subscriber.onNext(snapshot.toObject(FirestoreConferenceInfo::class.java))
-                }
+                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -94,13 +70,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
         return Observable.create { subscriber ->
             val registration = view(VIEW_SPEAKERS)
                 .collection(COLLECTION_SPEAKER_PAGES)
-                .addSnapshotListener { snapshot, exception ->
-                    if (exception != null && subscriber.isDisposed.not()) {
-                        subscriber.onError(exception)
-                        return@addSnapshotListener
-                    }
-                    subscriber.onNext(snapshot.documents.map { it.toObject(FirestoreSpeaker::class.java) })
-                }
+                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -111,13 +81,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
             val registration = view(VIEW_SPEAKERS)
                 .collection(COLLECTION_SPEAKER_PAGES)
                 .document(speakerId)
-                .addSnapshotListener { snapshot, exception ->
-                    if (exception != null && subscriber.isDisposed.not()) {
-                        subscriber.onError(exception)
-                        return@addSnapshotListener
-                    }
-                    subscriber.onNext(snapshot.toObject(FirestoreSpeaker::class.java))
-                }
+                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -127,13 +91,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
         return Observable.create { subscriber ->
             val registration = view(VIEW_EVENT_DETAILS)
                 .collection(COLLECTION_EVENTS)
-                .addSnapshotListener { snapshot, exception ->
-                    if (exception != null && subscriber.isDisposed.not()) {
-                        subscriber.onError(exception)
-                        return@addSnapshotListener
-                    }
-                    subscriber.onNext(snapshot.documents.map { it.toObject(FirestoreEvent::class.java) })
-                }
+                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -144,13 +102,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
             val registration = view(VIEW_EVENT_DETAILS)
                 .collection(COLLECTION_EVENTS)
                 .document(eventId)
-                .addSnapshotListener { snapshot, exception ->
-                    if (exception != null && subscriber.isDisposed.not()) {
-                        subscriber.onError(exception)
-                        return@addSnapshotListener
-                    }
-                    subscriber.onNext(snapshot.toObject(FirestoreEvent::class.java))
-                }
+                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -160,15 +112,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
         return Observable.create { subscriber ->
             val registration = view(VIEW_TRACKS)
                 .collection(COLLECTION_TRACKS)
-                .addSnapshotListener { snapshot, exception ->
-                    if (exception != null && subscriber.isDisposed.not()) {
-                        subscriber.onError(exception)
-                        return@addSnapshotListener
-                    }
-                    subscriber.onNext(snapshot.documents.map { trackSnapshot ->
-                        trackSnapshot.toObject(FirestoreTrack::class.java)
-                    })
-                }
+                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -181,15 +125,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
             val registration = db.collection(USER_DATA)
                 .document(userId)
                 .collection(COLLECTION_FAVORITES)
-                .addSnapshotListener { snapshot, exception ->
-                    if (exception != null && subscriber.isDisposed.not()) {
-                        subscriber.onError(exception)
-                        return@addSnapshotListener
-                    }
-                    subscriber.onNext(snapshot.documents.map { favoriteSnapshot ->
-                        favoriteSnapshot.toObject(FirestoreFavorite::class.java)
-                    })
-                }
+                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable(registration::remove)
         }
@@ -251,3 +187,42 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
         private const val DAY_DATE_SORTING = "day.date"
     }
 }
+
+@Suppress("TooGenericExceptionCaught") // Firestore doesn't have a strong Exception typing
+private inline fun <reified T> ObservableEmitter<List<T>>.injectSnapshotResult(snapshot: QuerySnapshot?, exception: Exception?) {
+    when {
+        exception != null && isDisposed.not() -> {
+            onError(exception)
+            return
+        }
+        else -> {
+            try {
+                onNext(snapshot?.toObjects(T::class.java)?.toList() ?: emptyList())
+            } catch (e: RuntimeException) {
+                onError(e)
+            }
+        }
+    }
+}
+
+@Suppress("TooGenericExceptionCaught") // Firestore doesn't have a strong Exception typing
+private inline fun <reified T> ObservableEmitter<T>.injectSnapshotResult(snapshot: DocumentSnapshot?, exception: Exception?) {
+    when {
+        exception != null && isDisposed.not() -> {
+            onError(exception)
+            return
+        }
+        snapshot?.exists() == true -> {
+            try {
+                onNext(snapshot.toObject(T::class.java))
+            } catch (e: RuntimeException) {
+                onError(e)
+            }
+        }
+        else -> onError(NonExistingDocumentException(snapshot))
+    }
+}
+
+private class NonExistingDocumentException(snapshot: DocumentSnapshot?) : RuntimeException(
+    "The specified path does not exist on the DB: ${snapshot?.reference?.path ?: "[null snapshot]"}"
+)

--- a/app/src/main/java/net/squanchy/service/firebase/FirestoreDbService.kt
+++ b/app/src/main/java/net/squanchy/service/firebase/FirestoreDbService.kt
@@ -25,7 +25,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
             val registration = view(VIEW_SCHEDULE)
                 .collection(COLLECTION_SCHEDULE_PAGES)
                 .orderBy(DAY_DATE_SORTING)
-                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
+                .addSnapshotListener { snapshot, exception -> subscriber.emitSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -36,7 +36,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
             val registration = db.collection(SOCIAL_STREAM)
                 .document(VIEW_TWITTER)
                 .collection(COLLECTION_TWEETS)
-                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
+                .addSnapshotListener { snapshot, exception -> subscriber.emitSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -50,7 +50,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
         return Observable.create { subscriber ->
             val registration = db.collection(CONFERENCE_INFO)
                 .document(VENUE)
-                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
+                .addSnapshotListener { snapshot, exception -> subscriber.emitSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -60,7 +60,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
         return Observable.create { subscriber ->
             val registration = db.collection(CONFERENCE_INFO)
                 .document(CONFERENCE)
-                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
+                .addSnapshotListener { snapshot, exception -> subscriber.emitSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -70,7 +70,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
         return Observable.create { subscriber ->
             val registration = view(VIEW_SPEAKERS)
                 .collection(COLLECTION_SPEAKER_PAGES)
-                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
+                .addSnapshotListener { snapshot, exception -> subscriber.emitSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -81,7 +81,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
             val registration = view(VIEW_SPEAKERS)
                 .collection(COLLECTION_SPEAKER_PAGES)
                 .document(speakerId)
-                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
+                .addSnapshotListener { snapshot, exception -> subscriber.emitSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -91,7 +91,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
         return Observable.create { subscriber ->
             val registration = view(VIEW_EVENT_DETAILS)
                 .collection(COLLECTION_EVENTS)
-                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
+                .addSnapshotListener { snapshot, exception -> subscriber.emitSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -102,7 +102,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
             val registration = view(VIEW_EVENT_DETAILS)
                 .collection(COLLECTION_EVENTS)
                 .document(eventId)
-                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
+                .addSnapshotListener { snapshot, exception -> subscriber.emitSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -112,7 +112,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
         return Observable.create { subscriber ->
             val registration = view(VIEW_TRACKS)
                 .collection(COLLECTION_TRACKS)
-                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
+                .addSnapshotListener { snapshot, exception -> subscriber.emitSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable { registration.remove() }
         }
@@ -125,7 +125,7 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
             val registration = db.collection(USER_DATA)
                 .document(userId)
                 .collection(COLLECTION_FAVORITES)
-                .addSnapshotListener { snapshot, exception -> subscriber.injectSnapshotResult(snapshot, exception) }
+                .addSnapshotListener { snapshot, exception -> subscriber.emitSnapshotResult(snapshot, exception) }
 
             subscriber.setCancellable(registration::remove)
         }
@@ -188,38 +188,46 @@ class FirestoreDbService(private val db: FirebaseFirestore) {
     }
 }
 
-@Suppress("TooGenericExceptionCaught") // Firestore doesn't have a strong Exception typing
-private inline fun <reified T> ObservableEmitter<List<T>>.injectSnapshotResult(snapshot: QuerySnapshot?, exception: Exception?) {
+private inline fun <reified T> ObservableEmitter<List<T>>.emitSnapshotResult(snapshot: QuerySnapshot?, exception: Exception?) {
     when {
-        exception != null && isDisposed.not() -> {
-            onError(exception)
-            return
-        }
-        else -> {
-            try {
-                onNext(snapshot?.toObjects(T::class.java)?.toList() ?: emptyList())
-            } catch (e: RuntimeException) {
-                onError(e)
-            }
-        }
+        exception != null && isDisposed.not() -> onError(exception)
+        else -> emitQueryResultValues(snapshot)
     }
 }
 
 @Suppress("TooGenericExceptionCaught") // Firestore doesn't have a strong Exception typing
-private inline fun <reified T> ObservableEmitter<T>.injectSnapshotResult(snapshot: DocumentSnapshot?, exception: Exception?) {
+private inline fun <reified T> ObservableEmitter<List<T>>.emitQueryResultValues(snapshot: QuerySnapshot?) {
+    try {
+        val modelItems = snapshot?.documents?.map { it.toObject(T::class.java) }
+            ?.filter { it != null }
+            ?: emptyList()
+
+        onNext(modelItems)
+    } catch (e: RuntimeException) {
+        onError(e)
+    }
+}
+
+private inline fun <reified T> ObservableEmitter<T>.emitSnapshotResult(snapshot: DocumentSnapshot?, exception: Exception?) {
     when {
-        exception != null && isDisposed.not() -> {
-            onError(exception)
-            return
-        }
-        snapshot?.exists() == true -> {
-            try {
-                onNext(snapshot.toObject(T::class.java))
-            } catch (e: RuntimeException) {
-                onError(e)
-            }
-        }
+        exception != null && isDisposed.not() -> onError(exception)
+        snapshot?.exists() == true -> emitExistingSnapshotValue(snapshot)
         else -> onError(NonExistingDocumentException(snapshot))
+    }
+}
+
+@Suppress("TooGenericExceptionCaught") // Firestore doesn't have a strong Exception typing
+private inline fun <reified T> ObservableEmitter<T>.emitExistingSnapshotValue(snapshot: DocumentSnapshot) {
+    try {
+        val modelObject = snapshot.toObject(T::class.java)
+
+        if (modelObject == null) {
+            onError(NonExistingDocumentException(snapshot))
+        } else {
+            onNext(modelObject)
+        }
+    } catch (e: RuntimeException) {
+        onError(e)
     }
 }
 

